File: packages/sdkwork-utils-csharp/I18nUtils.cs
using System.Globalization;

namespace Sdkwork.Utils;

public static class I18nUtils
{
    public static string FormatNumberLocale(double value, string locale, int decimals = 0)
    {
        var culture = locale.Equals("de-DE", StringComparison.OrdinalIgnoreCase)
            ? new CultureInfo("de-DE")
            : new CultureInfo("en-US");
        return value.ToString($"N{decimals}", culture);
    }

    public static string FormatDatetimeLocale(DateTimeOffset value, string locale)
    {
        var normalized = locale.ToLowerInvariant();
        var format = normalized.StartsWith("de") ? "dd.MM.yyyy HH:mm"
            : normalized.StartsWith("zh") ? "yyyy-MM-dd HH:mm"
            : "MM/dd/yyyy HH:mm";
        return value.ToUniversalTime().ToString(format, CultureInfo.InvariantCulture);
    }

    public static string? FormatDatetimeLocaleStr(string value, string locale)
    {
        var parsed = DateTimeUtils.ParseDatetime(value);
        return parsed is null ? null : FormatDatetimeLocale(parsed.Value, locale);
    }

    public static double? ParseNumberLocale(string input, string locale)
    {
        var trimmed = input.Trim();
        if (trimmed.Length == 0)
        {
            return null;
        }

        var decimalSeparator = locale.Equals("de-DE", StringComparison.OrdinalIgnoreCase) ? "," : ".";
        var groupingSeparator = locale.Equals("de-DE", StringComparison.OrdinalIgnoreCase) ? "." : ",";
        var normalized = trimmed.Replace(groupingSeparator, "").Replace(decimalSeparator, ".");
        return NumberUtils.ParseNumber(normalized);
    }
}

File: packages/sdkwork-utils-php/src/I18nUtils.php
<?php

declare(strict_types=1);

namespace Sdkwork\Utils;

final class I18nUtils
{
    public static function formatNumberLocale(float $value, string $locale, int $decimals = 0): string
    {
        return number_format($value, $decimals, $locale === 'de-DE' ? ',' : '.', $locale === 'de-DE' ? '.' : ',');
    }

    public static function formatDatetimeLocale(\DateTimeImmutable $value, string $locale): string
    {
        $utc = $value->setTimezone(new \DateTimeZone('UTC'));
        $normalized = strtolower($locale);
        if (str_starts_with($normalized, 'de')) {
            return $utc->format('d.m.Y H:i');
        }
        if (str_starts_with($normalized, 'zh')) {
            return $utc->format('Y-m-d H:i');
        }
        return $utc->format('m/d/Y H:i');
    }

    public static function formatDatetimeLocaleStr(string $value, string $locale): ?string
    {
        $parsed = DateTimeUtils::parseDatetime($value);
        return $parsed === null ? null : self::formatDatetimeLocale($parsed, $locale);
    }

    public static function parseNumberLocale(string $input, string $locale): ?float
    {
        $trimmed = trim($input);
        if ($trimmed === '') {
            return null;
        }
        $decimalSeparator = $locale === 'de-DE' ? ',' : '.';
        $groupingSeparator = $locale === 'de-DE' ? '.' : ',';
        $normalized = str_replace([$groupingSeparator, $decimalSeparator], ['', '.'], $trimmed);
        return NumberUtils::parseNumber($normalized);
    }
}
